### PR TITLE
Revert STS from 1.4 with cherry picks of integration and stress test fixes.

### PR DESCRIPTION
**Pull request by duderino**
This is an alternate to https://github.com/istio/envoy/pull/175.  It cherry picks some integration and stress test fixes from envoyproxy/envoy-wasm in addition to the STS reverts to try to get builds passing.

### CODE REVIEW

File: api/wasm/cpp/proxy_wasm_api.h [hunk 1]
```diff
@@ -355,9 +355,7 @@ class RootContext : public ContextBase {
   const std::string root_id_;
   std::unordered_map<uint32_t, HttpCallCallback> http_calls_;
   std::unordered_map<uint32_t, GrpcSimpleCallCallback> simple_grpc_calls_;
-  std::unique_ptr<GrpcCallHandlerBase> cur_grpc_call_;
   std::unordered_map<uint32_t, std::unique_ptr<GrpcCallHandlerBase>> grpc_calls_;
-  std::unique_ptr<GrpcStreamHandlerBase> cur_grpc_stream_;
```
Comment (PiotrSikora) [inline api/wasm/cpp/proxy_wasm_api.h]: You shouldn't be reverting those changes, see: https://github.com/istio/envoy/pull/175#issuecomment-594182783

Comment (duderino) [reply api/wasm/cpp/proxy_wasm_api.h]: But we didn't ship that in 1.4.6.  I think instead we should get release-1.4.6 into a state that looks like the release we just shipped, tag it, then reapply any fixes which can go out in 1.4.7

@howardjohn CC

Comment (PiotrSikora) [reply api/wasm/cpp/proxy_wasm_api.h]: That's a valid point. OK, fair enough, no one should be using Wasm in 1.4 anyway...

cc @mandarjog @kyessenov @bianpengyuan for visibility.

Comment (kyessenov) [reply api/wasm/cpp/proxy_wasm_api.h]: I disagree that we need to get release-1.4 branch into a state that looks like 1.4.6. We did not ship from 1.4 branch, as in reality we created a private 1.4.6 branch that was developed independently from 1.4 branch. Now that 1.4.6 branch is done, we should simply merge 1.4.6 branch into 1.4 branch (and that was exceptional due to secrecy around the patches).

**Comment by howardjohn** [comment]
> I disagree that we need to get release-1.4 branch into a state that looks
like 1.4.6

I agree. We need only to get release-1.4 branch into a state of what we
want to get shipped in 1.4.7.

**Comment by duderino** [comment]
OK, I can agree with that.  And John's point is accurate - we just need to get it into a state for what we want shipped in 1.4.7.

That means at this point that if there are pure bug fix SHAs they should stay in the branch.  So   @kyessenov @bianpengyuan send them in a PR or list the SHAs here and I'll get them back into the branch.   But if there are features mixed in with the SHAs, then we are still having a discussion with the community about whether we can include them in 1.4.7

**Comment by kyessenov** [comment]
I think the relevant stake holders are the release managers for 1.4, who have already looked into those PRs (in master or other branches sometimes and cherry-picked).
Please let's not call STS a google-specific extension. This is an IETF draft-ietf-oauth-token-exchange-16, which is implemented by AWS and other providers.

**Comment by duderino** [comment]
@kyessenov the issue is that these are features going into a patch release.  I am within my rights to object to that.

Also, I can't find a place where STS is called a google-specific extension.

**Comment by mandarjog** [comment]
@kyessenov 
1. We don’t have a clear view of what to do with vendor specific extensions yet. But now we will have to figure it out.
2. Since this is the Envoy repo,  if upstream envoy accepts something, we don’t need to have independent judgment about vendor specificity of the issue.

vendor specificity  is an issue for OSs istio-proxy repo, but we can set that aside for now. 

The primary issue then is feature addition in 1.4.5 of Istio. Since this code has never been shipped in a OSS release we can let it all roll back.

**Comment by kyessenov** [comment]
I get that features in a patch release is an issue, and I'm OK with pretending those changes never were approved. I just don't think we should be using the CVE release as the reason to revert those changes. It's making a bad precedent to have a bunch of changes reverted post-factum, and creates a lot of work figuring out which change corresponds to which feature. Clearly, we do want some important bug fixes in 1.4.7.

If I were to patch 1.4 six months from now (like what I had to do for 1.2), I'd be enormously confused by what happened in the commit history.

**Comment by bianpengyuan** [comment]
https://github.com/istio/envoy/pull/185 for the bug fix reverted by this pr.

**Pull request merged**